diff --git a/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/org/wso2/carbon/mediator/datamapper/ui/i18n/Resources.properties b/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/org/wso2/carbon/mediator/datamapper/ui/i18n/Resources.properties
index 3ab4e59..602e310 100644
--- a/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/org/wso2/carbon/mediator/datamapper/ui/i18n/Resources.properties
+++ b/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/org/wso2/carbon/mediator/datamapper/ui/i18n/Resources.properties
@@ -1,18 +1,19 @@
 mediator.datamapper.header=DataMapper Mediator
-mediator.datamapper.config=Config
-mediator.datamapper.inputSchema=InputSchema
-mediator.datamapper.outputSchema=OutputSchema
-mediator.datamapper.inputType=InputType
-mediator.datamapper.outputType=OutputType
+mediator.datamapper.configKey=Configuration
+mediator.datamapper.inputSchema=Input Schema
+mediator.datamapper.outputSchema=Output Schema
+mediator.datamapper.inputType=Input Type
+mediator.datamapper.outputType=Output Type
 
 mediator.datamapper.nsEditor=Name Space Editor
 mediator.datamapper.action=Action
 mediator.datamapper.addProperty=Add Property
 
 mediator.datamapper.delete=Delete
-mediator.datamapper.inputType.csv="text/csv"
-mediator.datamapper.inputType.xml="application/xml"
-mediator.datamapper.inputType.json="application/json"
-mediator.datamapper.outputType.csv="text/csv"
-mediator.datamapper.outputType.xml="application/xml"
-mediator.datamapper.outputType.json="application/json"
+mediator.datamapper.type.csv=text/csv
+mediator.datamapper.type.xml=application/xml
+mediator.datamapper.type.json=application/json
+
+
+mediator.datamapper.conf.registry.browser=Configuration Registry
+mediator.datamapper.gov.registry.browser=Governance Registry
diff --git a/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/web/datamapper-mediator/edit-mediator.jsp b/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/web/datamapper-mediator/edit-mediator.jsp
index dad889e..a2f7812 100644
--- a/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/web/datamapper-mediator/edit-mediator.jsp
+++ b/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/web/datamapper-mediator/edit-mediator.jsp
@@ -15,10 +15,11 @@
   --%>
 <%@ page language="java" contentType="text/html; charset=ISO-8859-1" pageEncoding="ISO-8859-1" %>
 <%@ page import="org.apache.axis2.context.ConfigurationContext" %>
+<%@ page import="org.apache.synapse.mediators.Value" %>
 <%@ page import="org.wso2.carbon.CarbonConstants" %>
 <%@ page import="org.wso2.carbon.ui.CarbonUIUtil" %>
 <%@ page import="org.wso2.carbon.utils.ServerConstants" %>
-<%@ page import="org.wso2.carbon.mediator.datamapper.DataMapperMediator" %>
+<%@ page import="org.wso2.carbon.mediator.datamapper.ui.DataMapperMediator" %>
 <%@ page import="org.wso2.carbon.mediator.service.ui.Mediator" %>
 <%@ page import="org.wso2.carbon.sequences.ui.util.SequenceEditorHelper" %>
 <%@ taglib uri="http://java.sun.com/jsp/jstl/fmt" prefix="fmt" %>
@@ -32,18 +33,30 @@
         throw new RuntimeException("Unable to edit the mediator");
     }
     DataMapperMediator dataMapperMediator = (DataMapperMediator) mediator;
-    
-    String config = "";
+
+    String configKeyVal = "";
+    String inputSchemaVal = "";
+    String outputSchemaVal = "";
+
+    Value configKey = null;
     if(dataMapperMediator.getConfigurationKey() != null) {
-        config = dataMapperMediator.getConfigurationKey();
+        System.out.println("getConfigurationKey is null");
+        configKey = dataMapperMediator.getConfigurationKey();
+        configKeyVal = configKey.getKeyValue();
     }
-	String inputSchema = "";
+
+	Value inputSchema = null;
     if(dataMapperMediator.getInputSchemaKey() != null) {
+        System.out.println("getInputSchemaKey is null");
         inputSchema = dataMapperMediator.getInputSchemaKey();
+        inputSchemaVal = inputSchema.getKeyValue();
     }
-	String outputSchema = "";
+
+	Value outputSchema = null;
     if(dataMapperMediator.getOutputSchemaKey() != null) {
+        System.out.println("getOutputSchemaKey is null");
         outputSchema = dataMapperMediator.getOutputSchemaKey();
+        outputSchemaVal = outputSchema.getKeyValue();
     }
 
     String url = CarbonUIUtil.getServerURL(this.getServletConfig().getServletContext(),
@@ -70,20 +83,20 @@
         <td>
             <table class="normal">
                 <tr>
-                    <td><fmt:message key="mediator.datamapper.config"/>
+                    <td><fmt:message key="mediator.datamapper.configKey"/>
                     </td>
                     <td>
-                        <input class="longInput" type="text" value="<%=config%>"
-                               name="config" id="config" readonly="true"/>
+                        <input class="longInput" type="text" value="<%=configKeyVal%>"
+                               name="configKey" id="configKey" readonly="true"/>
                     </td>
                     <td>
                         <a href="#registryBrowserLink"
                            class="registry-picker-icon-link"
-                           onclick="showRegistryBrowser('config','/_system/config')"><fmt:message
+                           onclick="showRegistryBrowser('configKey','/_system/config')"><fmt:message
                                 key="mediator.datamapper.conf.registry.browser"/></a>
                         <a href="#registryBrowserLink"
                            class="registry-picker-icon-link"
-                           onclick="showRegistryBrowser('config','/_system/governance')"><fmt:message
+                           onclick="showRegistryBrowser('configKey','/_system/governance')"><fmt:message
                                 key="mediator.datamapper.gov.registry.browser"/></a>
                     </td>
                 </tr>
@@ -91,7 +104,7 @@
                     <td><fmt:message key="mediator.datamapper.inputSchema"/>
                     </td>
                     <td>
-                        <input class="longInput" type="text" value="<%=inputSchema%>"
+                        <input class="longInput" type="text" value="<%=inputSchemaVal%>"
                                name="inputSchema" id="inputSchema" readonly="true"/>
                     </td>
                     <td>
@@ -109,7 +122,7 @@
                     <td><fmt:message key="mediator.datamapper.outputSchema"/>
                     </td>
                     <td>
-                        <input class="longInput" type="text" value="<%=outputSchema%>"
+                        <input class="longInput" type="text" value="<%=outputSchemaVal%>"
                                name="outputSchema" id="outputSchema" readonly="true"/>
                     </td>
                     <td>
@@ -128,23 +141,24 @@
                 <td>
                     <select id="mediator.datamapper.inputType" name="mediator.datamapper.inputType">
                         <%
+                            System.out.println("dataMapperMediator.getInputType()" + dataMapperMediator.getInputType());
                             if(dataMapperMediator.getInputType() == DataMapperMediator.CSV_VALUE){
                         %>
-                        <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.inputType.csv"/></option>
-                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.inputType.xml"/></option>
-                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.inputType.json"/></option>
+                        <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.type.csv"/></option>
+                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.type.xml"/></option>
+                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.type.json"/></option>
                         <%
                             } else if(dataMapperMediator.getInputType() == DataMapperMediator.XML_VALUE){
                         %>
-                         <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.inputType.csv"/></option>
-                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.inputType.xml"/></option>
-                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.inputType.json"/></option>
+                        <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.type.csv"/></option>
+                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.type.xml"/></option>
+                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.type.json"/></option>
                         <%
                             } else if (dataMapperMediator.getInputType() == DataMapperMediator.JSON_VALUE) {
                         %>
-                        <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.inputType.csv"/></option>
-                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.inputType.xml"/></option>
-                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.inputType.json"/></option>
+                        <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.type.csv"/></option>
+                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.type.xml"/></option>
+                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.type.json"/></option>
                         <%}%>
                     </select>
                 </td>
@@ -154,23 +168,24 @@
                 <td>
                     <select id="mediator.datamapper.outputType" name="mediator.datamapper.outputType">
                         <%
+                            System.out.println("dataMapperMediator.getOutputType()" + dataMapperMediator.getOutputType());
                             if(dataMapperMediator.getOutputType() == DataMapperMediator.CSV_VALUE){
                         %>
-                        <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.outputType.csv"/></option>
-                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.outputType.xml"/></option>
-                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.outputType.json"/></option>
+                        <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.type.csv"/></option>
+                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.type.xml"/></option>
+                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.type.json"/></option>
                         <%
                             } else if(dataMapperMediator.getOutputType() == DataMapperMediator.XML_VALUE){
                         %>
-                         <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.outputType.csv"/></option>
-                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.outputType.xml"/></option>
-                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.outputType.json"/></option>
+                         <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.type.csv"/></option>
+                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.type.xml"/></option>
+                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.type.json"/></option>
                         <%
                             } else if (dataMapperMediator.getOutputType() == DataMapperMediator.JSON_VALUE) {
                         %>
-                        <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.outputType.csv"/></option>
-                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.outputType.xml"/></option>
-                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.outputType.json"/></option>
+                        <option value="<%=DataMapperMediator.CSV_VALUE%>" selected="true"><fmt:message key="mediator.datamapper.type.csv"/></option>
+                        <option value="<%=DataMapperMediator.XML_VALUE%>"><fmt:message key="mediator.datamapper.type.xml"/></option>
+                        <option value="<%=DataMapperMediator.JSON_VALUE%>"><fmt:message key="mediator.datamapper.type.json"/></option>
                         <%}%>
                     </select>
                 </td>
diff --git a/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/web/datamapper-mediator/update-mediator.jsp b/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/web/datamapper-mediator/update-mediator.jsp
index 3b8a4bd..57eb8d1 100644
--- a/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/web/datamapper-mediator/update-mediator.jsp
+++ b/components/mediation-ui/mediators-ui/org.wso2.carbon.mediator.datamapper.ui/src/main/resources/web/datamapper-mediator/update-mediator.jsp
@@ -14,22 +14,11 @@
   ~  See the License for the specific language governing permissions and
   ~  limitations under the License.
   --%>
+<%@ page import="org.apache.synapse.mediators.Value" %>
 <%@ page import="org.wso2.carbon.mediator.service.ui.Mediator" %>
-<%@ page import="org.wso2.carbon.mediator.datamapper.DataMapperMediator" %>
+<%@ page import="org.wso2.carbon.mediator.datamapper.ui.DataMapperMediator" %>
 <%@ page import="org.wso2.carbon.sequences.ui.util.SequenceEditorHelper" %>
 
-<%! public boolean nullChecker(String strChecker) {
-
-    if (strChecker == null) {
-        return false;
-    } else if (strChecker != null) {
-        return true;
-    } else if ((!(strChecker.equalsIgnoreCase("")))) {
-        return true;
-    }
-    return false;
-}%>
-
 <%
     Mediator mediator = SequenceEditorHelper.getEditingMediator(request, session);
     if (!(mediator instanceof DataMapperMediator)) {
@@ -37,22 +26,26 @@
         throw new RuntimeException("Unable to edit the mediator");
     }
     DataMapperMediator dataMapperMediator = (DataMapperMediator) mediator;
-    String config = request.getParameter("config");
-    if(nullChecker(config)) {
-        dataMapperMediator.setConfigurationKey(config);
+
+    Value configKey = new Value(request.getParameter("configKey"));
+    if(configKey != null) {
+        System.out.println("configKey" + configKey + "|" + new Value(request.getParameter("configKey")));
+        dataMapperMediator.setConfigurationKey(configKey);
     }
-    String inputSchema = request.getParameter("inputSchema");
-    if(nullChecker(inputSchema)) {
+    Value inputSchema = new Value(request.getParameter("inputSchema"));
+    if(inputSchema != null) {
+        System.out.println("inputSchema" + configKey);
         dataMapperMediator.setInputSchemaKey(inputSchema);
     }
-    String outputSchema = request.getParameter("outputSchema");
-    if(nullChecker(outputSchema)) {
+    Value outputSchema = new Value(request.getParameter("outputSchema"));
+    if(outputSchema != null) {
+        System.out.println("inputSchema" + configKey);
         dataMapperMediator.setOutputSchemaKey(outputSchema);
     }
 
     dataMapperMediator.setInputType(Integer.parseInt(request.getParameter("mediator.datamapper.inputType")));
     dataMapperMediator.setOutputType(Integer.parseInt(request.getParameter("mediator.datamapper.outputType")));
-    dataMapperMediator.getProperties().clear(); // to avoid duplicates
+    //dataMapperMediator.getProperties().clear(); // to avoid duplicates
 
 %>
 
diff --git a/components/mediation-ui/mediators-ui/pom.xml b/components/mediation-ui/mediators-ui/pom.xml
index 525b82f..54cf07c 100644
--- a/components/mediation-ui/mediators-ui/pom.xml
+++ b/components/mediation-ui/mediators-ui/pom.xml
@@ -42,6 +42,7 @@
         <module>org.wso2.carbon.mediator.clone.ui</module>
         <module>org.wso2.carbon.mediator.command.ui</module>
         <module>org.wso2.carbon.mediator.conditionalrouter.ui</module>
+        <module>org.wso2.carbon.mediator.datamapper.ui</module>
         <module>org.wso2.carbon.mediator.dblookup.ui</module>
         <module>org.wso2.carbon.mediator.dbreport.ui</module>
         <module>org.wso2.carbon.mediator.drop.ui</module>
diff --git a/components/mediators/datamapper/org.wso2.carbon.mediator.datamapper/src/main/java/org/wso2/carbon/mediator/datamapper/DataMapperHelper.java b/components/mediators/datamapper/org.wso2.carbon.mediator.datamapper/src/main/java/org/wso2/carbon/mediator/datamapper/DataMapperHelper.java
index 62cb4f1..e5cd028 100644
--- a/components/mediators/datamapper/org.wso2.carbon.mediator.datamapper/src/main/java/org/wso2/carbon/mediator/datamapper/DataMapperHelper.java
+++ b/components/mediators/datamapper/org.wso2.carbon.mediator.datamapper/src/main/java/org/wso2/carbon/mediator/datamapper/DataMapperHelper.java
@@ -81,6 +81,7 @@ public class DataMapperHelper {
 		try {
 
 			// mapping resources needed to get the final output
+			//FIXME : remove caching part?? use in-memory for schemas
 			mappingResourceLoader = CacheResources.getCachedResources(context,
 					configkey, inSchemaKey, outSchemaKey, uuid);
 
@@ -154,8 +155,6 @@ public class DataMapperHelper {
 	 * 
 	 * @param inputDataType
 	 *            input data type
-	 * @param inMessage
-	 *            the incoming message
 	 * @return the input as a OMElement
 	 * @throws IOException
 	 */
diff --git a/components/mediators/datamapper/org.wso2.carbon.mediator.datamapper/src/main/java/org/wso2/carbon/mediator/datamapper/DataMapperMediator.java b/components/mediators/datamapper/org.wso2.carbon.mediator.datamapper/src/main/java/org/wso2/carbon/mediator/datamapper/DataMapperMediator.java
index ffe66b3..0411c81 100644
--- a/components/mediators/datamapper/org.wso2.carbon.mediator.datamapper/src/main/java/org/wso2/carbon/mediator/datamapper/DataMapperMediator.java
+++ b/components/mediators/datamapper/org.wso2.carbon.mediator.datamapper/src/main/java/org/wso2/carbon/mediator/datamapper/DataMapperMediator.java
@@ -39,6 +39,7 @@ import org.apache.synapse.mediators.Value;
  * input received by the previous mediator.
  */
 public class DataMapperMediator extends AbstractMediator implements ManagedLifecycle{
+//TODO : ManagedLifecycle why is this needed? what happens in init and destroy
 
 	private Value configurationKey = null;
 	private Value inputSchemaKey = null;
@@ -176,6 +177,7 @@ public class DataMapperMediator extends AbstractMediator implements ManagedLifec
 
 		boolean result = true;
 
+        //TODO : debug : what happens here?
 		String configkey = configurationKey.evaluateValue(messageContext);
 		String inSchemaKey = inputSchemaKey.evaluateValue(messageContext);
 		String outSchemaKey = outputSchemaKey.evaluateValue(messageContext);
@@ -189,14 +191,17 @@ public class DataMapperMediator extends AbstractMediator implements ManagedLifec
 		} else {
 			try {
 				// Does message conversion and gives the final result
+				//FIXME : move transform method to mediator class. remove helper class. remove synchronized
 				synchronized(DataMapperHelper.class){
 				DataMapperHelper.transform(messageContext, configkey, inSchemaKey,
 						outSchemaKey, inputType, outputType, getUniqueID());
 				}
 			} catch (SynapseException synExp) {
+                //FIXME : throw synapse exception
 				log.error("Mediation failed at DataMapperMediator");
 				result = false;
 			} catch (IOException e) {
+                //FIXME : throw synapse exception
 				log.error("Mediation failed at DataMapperMediator");
 				result = false;
 			}
@@ -232,6 +237,7 @@ public class DataMapperMediator extends AbstractMediator implements ManagedLifec
 	 */
 	@Override
 	public void destroy() {
+        //TODO :
 		if (id != null) {
 			setUniqueID(id);
         }		
diff --git a/components/mediators/datamapper/pom.xml b/components/mediators/datamapper/pom.xml
index 3f2dd26..2c9b782 100644
--- a/components/mediators/datamapper/pom.xml
+++ b/components/mediators/datamapper/pom.xml
@@ -1,16 +1,37 @@
-<?xml version="1.0" encoding="UTF-8"?>
-<project xmlns="http://maven.apache.org/POM/4.0.0"
-         xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
-         xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+<?xml version="1.0" encoding="utf-8"?>
+<!--
+ ~ Copyright (c) 2015, WSO2 Inc. (http://wso2.com) All Rights Reserved.
+ ~
+ ~ Licensed under the Apache License, Version 2.0 (the "License");
+ ~ you may not use this file except in compliance with the License.
+ ~ You may obtain a copy of the License at
+ ~
+ ~      http://www.apache.org/licenses/LICENSE-2.0
+ ~
+ ~ Unless required by applicable law or agreed to in writing, software
+ ~ distributed under the License is distributed on an "AS IS" BASIS,
+ ~ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ ~ See the License for the specific language governing permissions and
+ ~ limitations under the License.
+-->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
+
     <parent>
-        <artifactId>mediators</artifactId>
         <groupId>org.wso2.carbon.mediation</groupId>
+        <artifactId>mediators</artifactId>
         <version>4.6.1-BETA-SNAPSHOT</version>
-        <relativePath>../components/mediators/pom.xml</relativePath>
     </parent>
-    <modelVersion>4.0.0</modelVersion>
-
-    <artifactId>datamapper</artifactId>
 
+    <modelVersion>4.0.0</modelVersion>
+    <version>4.6.1-BETA-SNAPSHOT</version>
+    <artifactId>datamapper-mediator</artifactId>
+    <packaging>pom</packaging>
+    <name>WSO2 Carbon - DataMapper Mediator Module</name>
+    <url>http://wso2.org</url>
 
-</project>
\ No newline at end of file
+    <modules>
+        <module>org.wso2.carbon.mediator.datamapper.engine</module>
+        <module>org.wso2.carbon.mediator.datamapper</module>
+    </modules>
+</project>
diff --git a/components/mediators/pom.xml b/components/mediators/pom.xml
index d7e4a4c..574d4bc 100644
--- a/components/mediators/pom.xml
+++ b/components/mediators/pom.xml
@@ -41,5 +41,6 @@
         <module>oauth</module>
         <module>publishEvent</module>
         <module>cache</module>
+        <module>datamapper</module>
     </modules>
 </project>
diff --git a/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.feature/pom.xml b/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.feature/pom.xml
index 0d31130..5bf91ad 100644
--- a/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.feature/pom.xml
+++ b/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.feature/pom.xml
@@ -17,25 +17,25 @@
 <project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/maven-v4_0_0.xsd">
    <parent>
       <groupId>org.wso2.carbon.mediation</groupId>
-      <artifactId>cache-mediator-feature</artifactId>
+      <artifactId>datamapper-mediator-feature</artifactId>
       <version>4.6.1-BETA-SNAPSHOT</version>
    </parent>
    <modelVersion>4.0.0</modelVersion>
-   <artifactId>org.wso2.carbon.mediator.cache.feature</artifactId>
+   <artifactId>org.wso2.carbon.mediator.datamapper.feature</artifactId>
    <version>4.6.1-BETA-SNAPSHOT</version>
    <packaging>pom</packaging>
-   <name>WSO2 Carbon - Cache Mediator Aggregate Feature</name>
+   <name>WSO2 Carbon - DataMapper Mediator Aggregate Feature</name>
    <url>http://wso2.org</url>
-   <description>This feature contains the bundles required for Cache Mediator functionality</description>
+   <description>This feature contains the bundles required for DataMapper Mediator functionality</description>
    <dependencies>
       <dependency>
          <groupId>org.wso2.carbon.mediation</groupId>
-         <artifactId>org.wso2.carbon.mediator.cache.server.feature</artifactId>
+         <artifactId>org.wso2.carbon.mediator.datamapper.server.feature</artifactId>
          <type>zip</type>
       </dependency>
       <dependency>
          <groupId>org.wso2.carbon.mediation</groupId>
-         <artifactId>org.wso2.carbon.mediator.cache.ui.feature</artifactId>
+         <artifactId>org.wso2.carbon.mediator.datamapper.ui.feature</artifactId>
          <type>zip</type>
       </dependency>
    </dependencies>
@@ -53,14 +53,14 @@
                      <goal>p2-feature-gen</goal>
                   </goals>
                   <configuration>
-                     <id>org.wso2.carbon.mediator.cache</id>
+                     <id>org.wso2.carbon.mediator.datamapper</id>
                      <propertiesFile>../../../etc/feature.properties</propertiesFile>
                      <importFeatures>
                         <importFeatureDef>org.wso2.carbon.core:${carbon.kernel.version}</importFeatureDef>
                      </importFeatures>
                      <includedFeatures>
-                        <includedFeatureDef>org.wso2.carbon.mediation:org.wso2.carbon.mediator.cache.server.feature</includedFeatureDef>
-                        <includedFeatureDef>org.wso2.carbon.mediation:org.wso2.carbon.mediator.cache.ui.feature</includedFeatureDef>
+                        <includedFeatureDef>org.wso2.carbon.mediation:org.wso2.carbon.mediator.datamapper.server.feature</includedFeatureDef>
+                        <includedFeatureDef>org.wso2.carbon.mediation:org.wso2.carbon.mediator.datamapper.ui.feature</includedFeatureDef>
                      </includedFeatures>
                   </configuration>
                </execution>
diff --git a/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.server.feature/pom.xml b/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.server.feature/pom.xml
index d830139..f8e9e3e 100644
--- a/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.server.feature/pom.xml
+++ b/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.server.feature/pom.xml
@@ -19,22 +19,22 @@
 
   <parent>
       <groupId>org.wso2.carbon.mediation</groupId>
-      <artifactId>cache-mediator-feature</artifactId>
+      <artifactId>datamapper-mediator-feature</artifactId>
       <version>4.6.1-BETA-SNAPSHOT</version>
    </parent>
 
     <modelVersion>4.0.0</modelVersion>
-    <artifactId>org.wso2.carbon.mediator.cache.server.feature</artifactId>
+    <artifactId>org.wso2.carbon.mediator.datamapper.server.feature</artifactId>
    <version>4.6.1-BETA-SNAPSHOT</version>
     <packaging>pom</packaging>
-    <name>WSO2 Carbon - Cache Mediator  Feature</name>
+    <name>WSO2 Carbon - DataMapper Mediator  Feature</name>
     <url>http://wso2.org</url>
-    <description>This feature contains the Cache Mediator Backend</description>
+    <description>This feature contains the DataMapper Mediator Backend</description>
 
     <dependencies>
         <dependency>
             <groupId>org.wso2.carbon.mediation</groupId>
-            <artifactId>org.wso2.carbon.mediator.cache</artifactId>
+            <artifactId>org.wso2.carbon.mediator.datamapper</artifactId>
         </dependency>
     </dependencies>
 
@@ -52,7 +52,7 @@
                             <goal>p2-feature-gen</goal>
                         </goals>
                         <configuration>
-                            <id>org.wso2.carbon.mediator.cache.server</id>
+                            <id>org.wso2.carbon.mediator.datamapper.server</id>
                             <propertiesFile>../../../etc/feature.properties</propertiesFile>
                             <adviceFile>
                                 <properties>
@@ -61,7 +61,7 @@
                                 </properties>
                             </adviceFile>
                             <bundles>
-                                <bundleDef>org.wso2.carbon.mediation:org.wso2.carbon.mediator.cache</bundleDef>
+                                <bundleDef>org.wso2.carbon.mediation:org.wso2.carbon.mediator.datamapper</bundleDef>
                             </bundles>
                             <importFeatures>
                                 <importFeatureDef>org.apache.synapse.wso2:${carbon.mediation.version}</importFeatureDef>
diff --git a/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.ui.feature/pom.xml b/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.ui.feature/pom.xml
index fb68545..1d5bdf7 100644
--- a/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.ui.feature/pom.xml
+++ b/features/mediator-features/datamapper-mediator/org.wso2.carbon.mediator.datamapper.ui.feature/pom.xml
@@ -19,23 +19,23 @@
 
     <parent>
       <groupId>org.wso2.carbon.mediation</groupId>
-      <artifactId>cache-mediator-feature</artifactId>
+      <artifactId>datamapper-mediator-feature</artifactId>
       <version>4.6.1-BETA-SNAPSHOT</version>
     </parent>
 
 
     <modelVersion>4.0.0</modelVersion>
-    <artifactId>org.wso2.carbon.mediator.cache.ui.feature</artifactId>
+    <artifactId>org.wso2.carbon.mediator.datamapper.ui.feature</artifactId>
     <version>4.6.1-BETA-SNAPSHOT</version>
     <packaging>pom</packaging>
-    <name>WSO2 Carbon - Cache Mediator Management UI Feature</name>
+    <name>WSO2 Carbon - DataMapper Mediator Management UI Feature</name>
     <url>http://wso2.org</url>
-    <description>This feature contains the bundles required for Cache Mediator UI functionality</description>
+    <description>This feature contains the bundles required for DataMapper Mediator UI functionality</description>
 
     <dependencies>
         <dependency>
             <groupId>org.wso2.carbon.mediation</groupId>
-            <artifactId>org.wso2.carbon.mediator.cache.ui</artifactId>
+            <artifactId>org.wso2.carbon.mediator.datamapper.ui</artifactId>
         </dependency>
     </dependencies>
 
@@ -53,7 +53,7 @@
                             <goal>p2-feature-gen</goal>
                         </goals>
                         <configuration>
-                            <id>org.wso2.carbon.mediator.cache.ui</id>
+                            <id>org.wso2.carbon.mediator.datamapper.ui</id>
                             <propertiesFile>../../../etc/feature.properties</propertiesFile>
                             <adviceFile>
                                 <properties>
@@ -62,7 +62,7 @@
                                 </properties>
                             </adviceFile>
                             <bundles>
-                                <bundleDef>org.wso2.carbon.mediation:org.wso2.carbon.mediator.cache.ui</bundleDef>
+                                <bundleDef>org.wso2.carbon.mediation:org.wso2.carbon.mediator.datamapper.ui</bundleDef>
                             </bundles>
                             <importFeatures>
                                 <importFeatureDef>org.wso2.carbon.mediation.initializer.ui</importFeatureDef>
diff --git a/features/mediator-features/datamapper-mediator/pom.xml b/features/mediator-features/datamapper-mediator/pom.xml
index eafa5a8..a4ca1c4 100644
--- a/features/mediator-features/datamapper-mediator/pom.xml
+++ b/features/mediator-features/datamapper-mediator/pom.xml
@@ -22,13 +22,13 @@
    </parent>
    <modelVersion>4.0.0</modelVersion>
    <version>4.6.1-BETA-SNAPSHOT</version>
-   <artifactId>cache-mediator-feature</artifactId>
+   <artifactId>datamapper-mediator-feature</artifactId>
    <packaging>pom</packaging>
-   <name>WSO2 Carbon - Cache Mediator Feature Aggregator Module</name>
+   <name>WSO2 Carbon - DataMapper Mediator Feature Aggregator Module</name>
    <url>http://wso2.org</url>
    <modules>
-      <module>org.wso2.carbon.mediator.cache.server.feature</module>
-      <module>org.wso2.carbon.mediator.cache.ui.feature</module>
-      <module>org.wso2.carbon.mediator.cache.feature</module>
+      <module>org.wso2.carbon.mediator.datamapper.server.feature</module>
+      <module>org.wso2.carbon.mediator.datamapper.ui.feature</module>
+      <module>org.wso2.carbon.mediator.datamapper.feature</module>
    </modules>
 </project>
diff --git a/features/mediator-features/pom.xml b/features/mediator-features/pom.xml
index 399d7cb..b6f6856 100644
--- a/features/mediator-features/pom.xml
+++ b/features/mediator-features/pom.xml
@@ -40,6 +40,7 @@
         <module>org.wso2.carbon.mediators.ui.feature</module>
         <module>org.wso2.carbon.mediators.server.feature</module>
         <module>cache-mediator</module>
+        <module>datamapper-mediator</module>
     </modules>
 </project>
 
diff --git a/pom.xml b/pom.xml
index ea205ad..fb487c4 100644
--- a/pom.xml
+++ b/pom.xml
@@ -1288,6 +1288,11 @@
             </dependency>
             <dependency>
                 <groupId>org.wso2.carbon.mediation</groupId>
+                <artifactId>org.wso2.carbon.mediator.datamapper.ui</artifactId>
+                <version>${carbon.mediation.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>org.wso2.carbon.mediation</groupId>
                 <artifactId>org.wso2.carbon.mediator.callout.ui</artifactId>
                 <version>${carbon.mediation.version}</version>
             </dependency>
@@ -1894,6 +1899,24 @@
                 <version>${carbon.mediation.version}</version>
                 <type>zip</type>
             </dependency>
+            <dependency>
+                <groupId>org.wso2.carbon.mediation</groupId>
+                <artifactId>org.wso2.carbon.mediator.datamapper.feature</artifactId>
+                <version>${carbon.mediation.version}</version>
+                <type>zip</type>
+            </dependency>
+            <dependency>
+                <groupId>org.wso2.carbon.mediation</groupId>
+                <artifactId>org.wso2.carbon.mediator.datamapper.server.feature</artifactId>
+                <version>${carbon.mediation.version}</version>
+                <type>zip</type>
+            </dependency>
+            <dependency>
+                <groupId>org.wso2.carbon.mediation</groupId>
+                <artifactId>org.wso2.carbon.mediator.datamapper.ui.feature</artifactId>
+                <version>${carbon.mediation.version}</version>
+                <type>zip</type>
+            </dependency>
 
             <!--11 Previous version components -->
             <!--TODO use ndatasources-->
@@ -2243,6 +2266,12 @@
                 <artifactId>org.wso2.carbon.mediator.cache</artifactId>
                 <version>${carbon.mediation.version}</version>
             </dependency>
+            <dependency>
+                <groupId>org.wso2.carbon.mediation</groupId>
+                <artifactId>org.wso2.carbon.mediator.datamapper</artifactId>
+                <version>${carbon.mediation.version}</version>
+            </dependency>
+
 
             <!-- publishEvent mediator dependencies -->
             <dependency>
